crates/op-rbuilder/src/monitor_tx_pool.rs

```diff
@@ -1,68 +1,198 @@
-    match event {
-        FullTransactionEvent::Pending(hash) => {
-            info!(
-                target = "monitoring",
-                tx_hash = hash.to_string(),
-                kind = "pending",
-                "Transaction event received"
-            )
+impl<Pool> TransactionPoolMonitor<Pool>
+where
+    Pool: TransactionPool<Transaction = FBPooledTransaction> + Clone + 'static,
+{
+    pub fn rpc(&self) -> TransactionPoolMonitorRpc {
+        TransactionPoolMonitorRpc {
+            event_sender: self.event_sender.clone(),
         }
-        FullTransactionEvent::Queued(hash) => {
-            info!(
-                target = "monitoring",
-                tx_hash = hash.to_string(),
-                kind = "queued",
-                "Transaction event received"
-            )
+    }
+
+    pub async fn run(self) {
+        let mut new_transactions = self.pool.all_transactions_event_listener();
+
+        while let Some(event) = new_transactions.next().await {
+            // Push the event to the buffer
+            let event_data = TransactionEventData::from(event);
+            if self.log_events {
+                info!(
+                    target = "monitoring",
+                    tx_hash = event_data.hash.to_string(),
+                    kind = event_data.kind(),
+                    "Transaction event received"
+                )
+            }
+
+            if self.txpool_monitor {
+                println!("Sending event: {:?}", event_data);
+                let _ = self.event_sender.send(event_data);
+            }
         }
-        FullTransactionEvent::Mined {
-            tx_hash,
-            block_hash,
-        } => info!(
-            target = "monitoring",
-            tx_hash = tx_hash.to_string(),
-            kind = "mined",
-            block_hash = block_hash.to_string(),
-            "Transaction event received"
-        ),
-        FullTransactionEvent::Replaced {
-            transaction,
-            replaced_by,
-        } => info!(
-            target = "monitoring",
-            tx_hash = transaction.hash().to_string(),
-            kind = "replaced",
-            replaced_by = replaced_by.to_string(),
-            "Transaction event received"
-        ),
-        FullTransactionEvent::Discarded(hash) => {
-            info!(
-                target = "monitoring",
-                tx_hash = hash.to_string(),
-                kind = "discarded",
-                "Transaction event received"
-            )
+    }
+}
+
+pub struct TransactionPoolMonitorRpc {
+    event_sender: broadcast::Sender<TransactionEventData>,
+}
+
+#[async_trait]
+impl TxpoolExtApiServer for TransactionPoolMonitorRpc {
+    fn subscribe_events(
+        &self,
+        pending_subscription_sink: PendingSubscriptionSink,
+    ) -> SubscriptionResult {
+        println!("Subscribing to txpool events");
+        let mut event_receiver = self.event_sender.subscribe();
```

Comment:
If the buffer has elements, does this subscribe tries to read those elements first? or if will start at the beginning of the stream?